streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -1836,6 +1842,36 @@ public void maybeThrowTaskExceptionsFromProcessingThreads() {
         }
     }
 
+    private boolean transactionBuffersExceedCapacity = false;
+
+    boolean transactionBuffersExceedCapacity() {
+        return transactionBuffersExceedCapacity;
+    }
+
+    boolean transactionBuffersWillExceedCapacity() {
+        final boolean transactionBuffersAreUnbounded = maxUncommittedStateBytes < 0;
+        if (transactionBuffersAreUnbounded) {
+            return false;
+        }
+
+        // force an early commit if the uncommitted bytes exceeds or is *likely to exceed* the configured threshold
```

Comment:
I think this method would benefit from slightly more detailed docs/comments. I'm also a bit concerned about over-optimization, but lmk what you think:

First, IIUC, the idea is to use the number of additional uncommitted bytes since the last commit as a heuristic to estimate how many more uncommitted bytes will be added if we don't commit right now, and thus try to commit "just before" we go over the limit. Makes sense on its face, though it's obviously only as good as the buffer size increases are regular/linear. So the question is: will the buffer grow linearly between commits?

Unfortunately I suspect the answer is "no" in many cases -- In fact, the only case in which the buffer will grow by the same amount each iteration is when every insert is a new key. Of course, most KV stores have a bounded keyspace (else they'd grow forever) so we expect to see updates to the same keys. The very first iteration after a commit will show the most growth, and that will taper off as we rebuild the keyset and see fewer and fewer new keys vs in-place updates.

Of course it's not incorrect to commit early, I just wanted to get your thoughts on this. Maybe a good middle ground would be to maintain a running average `deltaBytes`? Or even just cut the previous `deltaBytes` in half? Either way it's still just an approximation, and things like maxUncommittedBytes are always just best-effort in the end.

Reply:
> First, IIUC, the idea is to use the number of additional uncommitted bytes since the last commit as a heuristic to estimate how many more uncommitted bytes will be added if we don't commit right now, and thus try to commit "just before" we go over the limit.

Correct, this is the intent. This behaviour was requested on the mailing list, IIRC by @cadonna, to reduce the risk of OOM between commits. The thinking was that users will  expect `statestore.uncommitted.max.bytes` to be an upper-bound, and likely set it with little headroom. If the commit interval is fairly long (e.g. 30 seconds under ALOS), then there's a good chance we would exceed this limit by a considerable margin.

> [...] so we expect to see updates to the same keys. [...] and see fewer and fewer new keys vs in-place updates.

Writes in RocksDB are the same size, irrespective of whether they are inserts of new keys, or updates to existing keys. The only case I can think of where this may be different is when migrating an existing record from a non-timestamped to a timestamped column family; in which case, the updates would actually be slightly larger, due to the additional tombstone written.

Ultimately, any heuristic we choose will be imperfect. If we track an average, even since the last commit, then it would be less responsive to sudden increases in throughput, potentially allowing an OOM before the average has had a chance to catch up. The current approach attempts to err on the side of caution: we may trigger a commit unnecessarily early on occasion, but we should be less likely to exceed the configured buffer size.

I'm not against changing this heuristic to a running average, or anything else; I'm just not sufficiently convinced it would be better :grin:

Reply:
> Writes in RocksDB are the same size, irrespective of whether they are inserts of new keys, or updates to existing keys.

I could be wrong about this, but I thought that writing to an existing key in the active memtable -- ie before it is flushed and/or made immutable  --  is indeed applied as an overwrite, and does not grow the write buffer's overall size? Of course this only applies to patterns in which keys are updated multiple times between flushes, sure. Still, I feel like this is one of the major points of the write buffer: to deduplicate same-key writes as much as possible before flushing them, since once they're flushed it has to wait for a compaction to deduplicate further.

That said, I don't feel strongly about this, and any potential performance degradation should come to light in the benchmarks. Thanks for explaining your thoughts on this